[WIP] Adding a new message

hamishwillee opened the pull request:
1. This "accidentally" includes the file format PR that may not be included. That means I will probably delete and recreate this in future. I am however going to create this now as a working area for discussion.

hamishwillee commented:
#### Questions 1: Versions

It looks in previous docs like a dialect takes the version of its included files. 
```
<!-- NOTE: If the included file already contains a version tag, remove the version tag here, else uncomment to enable. -->
    <!-- <version>6</version> -->
```
That does not make sense because you can include multiple dialect files - so which would be used.

1. Is this correct?
2. If you have multiple includes which have a version, which will be used
3. If you put in a version in a dialect what will be used.
4. Is the version field used anywhere in code? if so, what for?
5. Following on from this, when should the version number iterate.

hamishwillee commented:
#### Question 2: Dialect

```
<dialect>8</dialect>
```
1. What is this used for? Ie by generator or by code? 
2. Can you query the dialect a message came from
3. What if you reuse one?
4. What number/range can people use for this - just "check the existing values in the directory and use one that isn't allocated?

hamishwillee commented:
#### Question 3: Including

I have said you can include multiple files, that they can be at any file location (ie relative) and that they can also include other files. I have also advised that having relative paths and multi-levels of linking are not tested.

Is that correct? It appears to be.

hamishwillee commented:
#### Question 4: Ranges for messages

What range should a dialect use for its messages in MAVLink 2?

hamishwillee commented:
#### Question 5 Ranges for Enums

1. I assume that if an enum is overloaded by a dialect it will be appended to dialect definition?
2. What enum range should dialect use?
3. What if they overload common enum of same ID
4. Are there any issues with multiple dialects using the same range?

LorenzMeier commented:
Best discussed on a call

amilcarlucas commented on en/guide/create_message.md:
```diff
@@ -0,0 +1,246 @@
+      - All numbers below 255 should be considered reserved unless messages are also intended for MAVLink 1. 
+        > **Note** IDs are precious in MAVLink 1!
+  - `name`: The name attribute provides a human readable form for the message (ie "BATTERY_STATUS"). It is used for naming helper functions in generated libraries, but is not sent over the wire.
+- `description` (optional): Human readable description of message, shown in user interfaces and in code comments. 
```
I'd vote that we should change this field to *NOT* be optional!

hamishwillee replied:
@amilcarlucas I agree. I created a PR for discussion of this. https://github.com/ArduPilot/pymavlink/pull/215

amilcarlucas commented on en/guide/create_message.md:
```diff
@@ -0,0 +1,246 @@
+  This should contain all information (and hyperlinks) to fully understand the message.
+- `field`: Encodes one field of the message. The field value is its name/text string used in GUI documentation (but not sent over the wire).
+  - `type`: Similar to a field in a C struct - the size of the data required to store/represent the data type.
+    - Fields can be signed/unsigned integers of size 8, 16, 23, 64 bits (`{u)int8_t`, `(u)int16_t`, `(u)int32_t`, `(u)int64_t`), single/double precision IEEE754 floating point numbers.They can also be arrays of the other types - e.g. `uint16_t[10]`. 
```
White space missing after floating point numbers.

amilcarlucas commented on en/guide/create_message.md:
```diff
@@ -0,0 +1,246 @@
+    - Fields can be signed/unsigned integers of size 8, 16, 23, 64 bits (`{u)int8_t`, `(u)int16_t`, `(u)int32_t`, `(u)int64_t`), single/double precision IEEE754 floating point numbers.They can also be arrays of the other types - e.g. `uint16_t[10]`. 
+  - `name`: Name of the field (used in code).
+  - `enum`: Name of an enum defining possible values of the field (e.g. `MAV_BATTERY_CHARGE_STATE`).
+  - `units`: The units for fields that take numeric values (not enums). These are defined in the [schema](https://github.com/ArduPilot/pymavlink/blob/master/generator/mavschema.xsd) (search on *name="SI_Unit"*)
```
might be a good idea to also document `print_format` and `display` attributes.
There is also an attribute called `default`that seams to not be used anywhere. That one should maybe removed from the spec. @LorenzMeier what do you think?

hamishwillee commented:
OK, I have updated this in light of feedback
- fixed typos
- added actual behaviour for version usage
- added actual behaviour for include.

Still to do:
1. Enum docs
2. MAV_CMD docs
3. Open issues above.
4. Most important, above covers "how" and "where" to create messages, but not best practise - ie "How do I design a sensible message", "How do I go about getting a message discussed".

@LorenzMeier @amilcarlucas I'd like to get this in "more or less" as is as a starting point. Can we merge and then continue?

amilcarlucas commented:
I'm all in favor of merging this, it is a leap forward!

hamishwillee commented:
Thanks @amilcarlucas , I agree. I'll give @LorenzMeier a few days to comment. If not comment I'll merge in at the point just before things become "opinion".

Today I:
- Split out the info about XML file format into its own doc - essentially a human readable form of the real schema
- To schema doc I added a section on enums. Would appreciate your re-review of that doc. It is essentially done.

The "How to create a new message" is what was left. I will extend this with more general information about writing enums, messages and mav_cmds. It effectively will be more process and style than specific how to. I think that works.

amilcarlucas commented on en/guide/message_schema.md:
```diff
@@ -0,0 +1,248 @@
+The `deprecated` attributes are:
+* `since`: Year/month when deprecation started. Format: YYYY-MM.
+* `replaced by`: String of entity that supersedes this item.
+* `description` (optional): String with more information about the deprecation.
```
description is not an attribute, but an element, hence it is inside the <deprecated>  xxx <deprecated/>

amilcarlucas commented on en/guide/message_schema.md:
````diff
@@ -0,0 +1,248 @@
+<wip />
+```
+
+The `wip` tags are:
````
s/are/can contain an optional description text/

amilcarlucas commented on en/guide/message_schema.md:
```diff
@@ -0,0 +1,248 @@
+  - The tag should be used for MAVLink 1 messages only (id < 256) that have been extended in MAVLink 2. 
+
+
+### Enum Definition (enums) {#enum}
```
Move the enum above the messages

hamishwillee replied:
Ok.

hamishwillee commented:
Thanks @amilcarlucas - fixed those issues in https://github.com/mavlink/mavlink-devguide/pull/96/commits/82be2790ceeb3639638b64ce0cecf10e67397e13

hamishwillee commented on en/guide/message_schema.md:
````diff
@@ -0,0 +1,231 @@
+<wip />
+```
+
+<!--  ref="description" minOccurs="0" -->
````
@amilcarlucas Across the schema I see this a number of times. Of course ideally it should be one, but should it not be restricted everywhere to max of one?

hamishwillee commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,205 @@
+
+TBD
+
+<!-- 
```
@amilcarlucas I've made some questions/assumptions in this section and in Messages. If you happen to know answers to the questions, please add notes :-)

hamishwillee commented:
OK, so you can see further changes. Essentially this is starting to flesh out the "Defining XML elements" section.

amilcarlucas commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,205 @@
+<!-- 
+Assumptions
+- You can add new enums without risk provided they have a unique name
+- You can add new entries to an enum provided each has a unique name and value (what if clash)
```
I'm not sure if pymavlink tests if values clash. That needs to be tested.

amilcarlucas commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,205 @@
+- You can add new enums without risk provided they have a unique name
+- You can add new entries to an enum provided each has a unique name and value (what if clash)
+- You can add new entries for an existing enum in dialect, provided again that the value/name is unique
+- entry values and numbers must be unique ? - 
```
Again, I'm not sure if pymavlink will test this. And there might be other generators that do not test it. So yes, please tell the users to take care.

amilcarlucas commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,205 @@
+
+Questions
+- If a dialect adds enums, what ranges can it use?
+
```
The values must be in the range of all the message fields 'type' that use that enum. Again, I'm not sure if pymavlink tests this.

hamishwillee replied:
I mean that each enum entry has an id. If you want to extend an ID in a way that does not clash with the original what ids could you use. In most cases there is no way to proscribe this, but you'd need to do so for MAV_CMDs.

amilcarlucas commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,205 @@
+  - Q: If a dialect adds new messages, 
+      - what id ranges can it use in mavlink 1 and mavlink 2
+      - What is allocation strategy?
+- There is a maximum size for a message/max number of fields?
```
For mavlink 1.0 I think the entire message must be smaller than 256 bytes.

hamishwillee replied:
OK, so FYI, the payload max size is 255 bytes in both versions (message size is bigger than that). However the generator does not appear to check that.
The maximum number of fields is 64. The generator crashes with an exception for any more.

amilcarlucas commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,205 @@
+- You can add a new field to MAVLink 1 message if you prefix with extensions. Field must have a unique name.
+- You can't add a new field to a MAVLink 2 range message without breaking compatibility due to field reordering. (Check)
+- new fields must be added in the file that declares message. You can't add additional fields in a dialect.
+- Use units and type. Should or must have a description?
```
I think we should make description mandatory. Units should be required, but not mandatory.
e.g. no longer allow accelerometer raw data fields without units, all fields *must* have units, unless unitless.

amilcarlucas commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,205 @@
+- new fields must be added in the file that declares message. You can't add additional fields in a dialect.
+- Use units and type. Should or must have a description?
+- Any advice on design of messages? - "What questions should you ask to test a new message"
+  - Each field should only have one unit - ie no changing units based on some other value.
```
YES, Please

hamishwillee commented on en/guide/define_xml_element.md:
````diff
@@ -0,0 +1,336 @@
+```
+
+
+### Creating a Message
````
@amilcarlucas Just FYI, this is still work in progress, but I have updated this by inferring the answers to some of my questions. I'll continue with enums tomorrow.

Basically you can infer that there are allocated ranges for message ids in MAVLink 2. I have suggested creating an issue to request a new range.

I also tested what mavgen reports errors against.

amilcarlucas replied:
Nice !

hamishwillee commented:
I also sent this query to the MAVLink discussion group - @amilcarlucas , others, any thoughts/confirmations? ....

My hunch is that MAV_CMDs were originally intended for mission commands
- At some point someone realised that it would be useful to use the same command outside of missions - which resulted in the Command protocol. 
- Then in MAVLink 1 message IDs started to run out, so we started using the command protocol and the much bigger enum id range to send commands which we'd otherwise have used normal messages for.

So,
1. Are the MAV_CMDs = Mission Commands or MAVLink commands?
2. When should a command be written as a MAV_CMD and when as a stand-alone message? 

Can we give any recommendations?

amilcarlucas commented:
My guess is that if it needs a limited amount of parameters (7) and/or needs a single lat/lon/alt point and less than 4 parameters : use MAV_CMD

If it needs more than 7 parameters, use a proper message.

Currently the advantage of messages is that they are better documented, but once ardupilot/pymavlink#211 gets merged, that can change ;)

hamishwillee commented:
Thanks @amilcarlucas 
Those are good reasons :-). Michael also came up with another one - the command protocol has a standardised way to handle ACK/NACK - see https://groups.google.com/forum/#!topic/mavlink/V67JJqX1iQQ

I'll update the docs inline.

hamishwillee commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,424 @@
+
+> **Note** For detailed information about the file format see [MAVLink XML Schema](../guide/xml_schema.md#messages) (you can also inspect [common.xml](https://github.com/mavlink/mavlink/tree/master/message_definitions/v1.0/common.xml) and other dialect files). 
+
+## Should we Define Messages or Commands?
```
FYI @amilcarlucas , new info about commands vs messages. I added to top here as it is possibly first decision someone should consider.

hamishwillee commented:
@amilcarlucas I think this is now complete as far as "information needed" goes. There is one open question https://github.com/mavlink/mavlink-devguide/pull/96#issuecomment-409112642 about what dialect number is for and how allocated, but that is not a blocker, since it really does not appear to be used.

I will review this again next week. My current feeling is that it is just about right. However it may be that this can be better structured to reduce the duplication between the schema doc and the defining an element doc. Also to find a better "home" for where these docs live - ie some grouping that isn't just "Guide".

thomasgubler commented on en/guide/define_xml_element.md:
```diff
@@ -0,0 +1,424 @@
+  (MAVLink 1 has a much larger free pool of ids for MAVLink commands than for message ids).
+* Your command/message requires an ACK/NACK; using the existing protocol acknowledgments may be faster/easier than defining another message for acknowledgments.
+
+Otherwise either method may freely be used.
```
Leaving size/fitting constraints aside I always thought of it this way:
A **command** is used to for sporadic events where you want to trigger something. It is important that the the command is not missed, hence the ack.
A **message** is used to stream a continuous flow of data (might be high or low frequency, but it is not a one time event). Some data might get lost but the receiving side will deal with that.

hamishwillee replied:
Thanks @thomasgubler 

Hmmm, the actual division of existing messages/commands is not that clean. Protocols often provide acks and use messages directly (all the command protocol does for you is formalise that).  I guess you must use messages and an ack-less protocol for streams. 
You do need to use messages for broadcast streams, because command protocol assumes a destination.

Upshot, I think it is reasonable that these be added as additional reasons to each list above, rather than replacing them.

Does that make sense?

amilcarlucas replied:
yes

hamishwillee replied:
Cool. Update added.

hamishwillee commented:
@LorenzMeier Should be good to review.
